comit/src/expiries.rs

```diff
@@ -0,0 +1,631 @@
+        let beta_start_time = beta_now.add_duration(ttl);
+        let beta_expiry = beta_start_time.add_duration(alice_needs);
+
+        // Bob redeems on alpha ledger so needs time to act before the alpha expiry.
+        let bob_needs = period_for_bob_to_complete(&config, BobState::Created);
+        let alpha_now = alpha.current_time();
+        let alpha_start_time = alpha_now.add_duration(ttl);
+        let alpha_expiry = alpha_start_time.add_duration(bob_needs);
+
+        Expiries {
+            config,
+            alpha_connector: alpha,
+            beta_connector: beta,
+            alpha_expiry,
+            beta_expiry,
+        }
+    }
+
+    /// We define 'useful' to mean:
+    /// 1. Alice has time to transition from current state to `Done`.
+    /// 2. Bob has time to transition from current state to `Done`.
+    ///
+    /// An actor only knows there own state but can reason about the state of
+    /// the counterparty.
+    // BUG: Assuming an actor uses their own state to calculate usefulness to
+    // themselves and their guess at the couterparties state to calculate
+    // usefulness to the counterparty we can find ourselves in a position where
+    // we continue because the swap is useful to us but the counterparty
+    // immediately cancels because they believe the swap to be not-useful to us.
+    // This is due to the fact that we must be conservative in our guess of the
+    // counterparties state, hence we may include additional unnecessary
+    // transitions in our calculation. This will only become an issue as we approach
+    // the time cutoff for usefulness i.e., while actors act comfortably within
+    // their 'time to act' window we should not hit this bug.
+    pub fn is_useful<S>(&self, current_state: S) -> bool
+    where
+        S: AsAliceState + AsBobState,
+    {
+        let bob_state = current_state.as_bob_state();
+        let alice_state = current_state.as_alice_state();
+
+        self.alice_has_time_to_complete(alice_state) && self.bob_has_time_to_complete(bob_state)
+    }
+
+    /// True if Alice has time to complete a swap (i.e. transition to done).
+    fn alice_has_time_to_complete(&self, current_state: AliceState) -> bool {
+        let period = period_for_alice_to_complete(&self.config, current_state);
+        let now = self.beta_connector.current_time();
+
+        // Alice redeems on beta ledger so is concerned about the beta expiry.
+        let end_time = now.add_duration(period);
+        end_time < self.beta_expiry
+    }
+
+    /// True if Bob has time to complete a swap (i.e transition to done).
+    fn bob_has_time_to_complete(&self, current_state: BobState) -> bool {
+        let period = period_for_bob_to_complete(&self.config, current_state);
+        let now = self.alpha_connector.current_time();
+
+        // Bob redeems on alpha ledger so is concerned about the alpha expiry.
+        let end_time = now.add_duration(period);
+        end_time < self.alpha_expiry
```

Comment:
Thanks for looking at this @da-kami.

>  Does this take `Refund` into account? (I assume yes, then it is correct.)

This calculates if Bob has time to complete the happy path only, does not take refund into account.

> I am a bit confused about what the state `Done` stands for, because as far as I saw it does not take `Refund` into account. I guess because if depicts the happy-path. If we cannot reach `Done` we either decide to not fund at all or we go for `Refund` if we already funded as Bob.

Correct, done is happy path complete i.e., 'swapped'.

> Out of the context, I wonder if above condition is correct:

I believe the condition is correct because it is determining whether Bob has time to complete all his actions hence it is against alpha expiry.

> Bob can only redeem till `beta_expiry` is reached! After that he has to go for refund. The time window between `beta_expiry` and `alpha_expiry` is his refund window. (Alice can refund after `alpha_expiry` is reached, she is protected by the secret, Bob is not.)

You are correct in that as soon as beta expiry is available Bob should take it, this logic is within `next_action_for_bob` in this module. The check on expiry is done _before_ we call the code mentioned above. Oh you inadvertently found a bug for me, I was checking the wrong expiry in both the `next_action_for_...` methods!  

> I think it should hence be `beta_expiry` here, if you only talk about `Redeeming` in `Done`.
> 
> Please correct me if I have a thinking mistake :)

Reply:
Man, on further thought you might be onto something here. Even if Bob has time to act before his alpha expiry the beta expiry overrides that. Will think more about this, thanks.

Reply:
Yes, it should be correct. I took a look at [the HTLC expiry documentation](https://comit.network/docs/core-concepts/atomic-swap-htlc/#htlc-expiries) and I think my statement was actually not completely correct.

Bob has time till `alpha_expiry` is reached to redeem or refund. Alice has to redeem until beta expiry is reached. But Bob can actually redeem even after beta expiry is reached as long as he has still enough time to get his transaction confirmed before `alpha-expiry` is reached.

Reply:
> Nice work!
> I add the `no-mergify` label just in case, seems you are still working in review comments.

Cool thanks, I forgot that :)